routes/artifacts.js

```diff
@@ -0,0 +1,597 @@
+    "",
+    "**Artifact immutability**, generally speaking you cannot overwrite an",
+    "artifact when created. But if you repeat the request with the same",
+    "properties the request will succeed as the operation is idempotent.",
+    "This is useful if you need to refresh a signed URL while uploading.",
+    "Do not abuse this to overwrite artifacts created by another entity!",
+    "Such as worker-host overwriting artifact created by worker-code.",
+    "",
+    "As a special case the `url` property on _reference artifacts_ can be",
+    "updated. You should only use this to update the `url` property for",
+    "reference artifacts your process has created."
+  ].join('\n')
+}, async function(req ,res) {
+  // Validate parameters
+  if (!api.checkParams(req, res)) {
+    return;
+  }
+
+  var taskId      = req.params.taskId;
+  var runId       = parseInt(req.params.runId);
+  var name        = req.params.name;
+  var input       = req.body;
+  var storageType = input.storageType;
+  var contentType = input.contentType || 'application/json';
+
+  // Find expiration date
+  var expires = new Date(input.expires);
+
+  // Validate expires it is in the future
+  var past = new Date();
+  past.setMinutes(past.getMinutes() - 15);
+  if (expires.getTime() < past.getTime()) {
+    return res.status(400).json({
+      message:  "Expires must be in the future",
+      error: {
+        now:      new Date().toJSON(),
+        expires:  expires.toJSON()
+      }
+    });
+  }
+
+  // Load Task entity
+  var task = await this.Task.load({
+    taskId:       taskId
+  }, true);
+
+  // Handle cases where the task doesn't exist
+  if (!task) {
+    return res.status(404).json({
+      message: "Task not found"
+    });
+  }
+
+  // Check presence of the run
+  var run = task.runs[runId];
+  if (!run) {
+    return res.status(404).json({
+      message: "Run not found"
+    });
+  }
+
+  // Get workerGroup and workerId
+  var workerGroup = run.workerGroup;
+  var workerId    = run.workerId;
+
+  // Authenticate request by providing parameters
+  if(!req.satisfies({
+    workerGroup:  workerGroup,
+    workerId:     workerId,
+    name:         name
+  })) {
+    return;
+  }
+
+  // Validate expires <= task.expires
+  if (expires.getTime() > task.expires.getTime()) {
+    return res.status(400).json({
+      messages: "Artifact expires before the task (task.expires < expires)",
+      error: {
+        taskExpires:      task.expires.toJSON(),
+        expires:          expires.toJSON()
+      }
+    });
+  }
+
+  // Ensure that the run is running
+  if (run.state !== 'running') {
+    return res.status(409).json({
+      message:    "The given is not running",
+      error: {
+        status:   task.status(),
+        runId:    runId
+      }
+    });
+  }
+
+  // Construct details for different storage types
+  var isPublic = /^public\//.test(name);
+  var details  = {};
+  if (storageType === 's3') {
```

Comment:
Okay, I couldn't help not trying this.. But one could also do:
```js
// with switch statement 
switch(storageType) {
    case 's3':
      if (isPublic) {
        details.bucket  = this.publicBucket.bucket;
      } else {
        details.bucket  = this.privateBucket.bucket;
      }
      details.prefix    = [taskId, runId, name].join('/');
      break;

    case 'azure':
      details.container = this.blobStore.container;
      details.path      = [taskId, runId, name].join('/');
      break;

    case 'reference':
      details.url       = input.url;
      break;

    case 'error':
      details.message   = input.message;
      details.reason    = input.reason;
      break;

    default:
      throw new Error("Unknown storageType: " + storageType);
  }


/// With es6 dictionary:
{
    s3() {
      if (isPublic) {
        details.bucket  = this.publicBucket.bucket;
      } else {
        details.bucket  = this.privateBucket.bucket;
      }
      details.prefix    = [taskId, runId, name].join('/');
    }
    azure() {
      details.container = this.blobStore.container;
      details.path      = [taskId, runId, name].join('/');
    }
    reference() {
      details.url       = input.url;
    }
    error() {
      details.message   = input.message;
      details.reason    = input.reason;
    }
  }[storageType]();


// If writing a custom utility function called Switch implemented as
var Switch = (choice, options) => { return options[choice](); };
Switch(storageType, {
    s3() {
      if (isPublic) {
        details.bucket  = this.publicBucket.bucket;
      } else {
        details.bucket  = this.privateBucket.bucket;
      }
      details.prefix    = [taskId, runId, name].join('/');
    }
    azure() {
      details.container = this.blobStore.container;
      details.path      = [taskId, runId, name].join('/');
    }
    reference() {
      details.url       = input.url;
    }
    error() {
      details.message   = input.message;
      details.reason    = input.reason;
    }
  });
```

I love ES6 :)
Granted I went with switch statement, it's the decent thing to do here :)